### ISSUE TITLE
[rl] Grug attention materializes seq x seq: context ceiling ~8k vs declared 65k; why is flash-attn rejected?

### ISSUE THREAD

**Issue by penfever** [marin-community/MarinSkyRL#232]
Grug's attention materializes the full `[batch, heads, seq, seq]` score matrix in FP32 and applies the sliding window as a mask rather than as a banded computation. `_supports_sdpa = False` and `_supports_flash_attn = False` (`grug_moe.py:508-509`) leave no memory-efficient fallback, so **training context is capped around 8k** against a checkpoint that declares `max_position_embeddings: 65536`.

This is not a geometry problem — no node count fixes it.

## The ceiling

Config: 26 layers, 20 attention heads (5 KV heads `repeat_interleave`d to 20), `head_dim` 128, `sliding_window` 2048.

The FP32 score tensor is `20 × s² × 4 B = 80·s²` per layer, and the forward holds roughly three concurrently — the einsum output, the `torch.where` result, and the softmax output before its BF16 cast. Gradient checkpointing defaults on (`ppo_base_config.yaml:205`), so only one layer is live during recompute and the peak is per-layer, not ×26.

At ~240·s² bytes, batch 1: **4k → 4.0 GB · 8k → 16.1 GB · 16k → 64.4 GB · 65k → 1,031 GB.** Against the ~20–30 GB an 80 GB H100 has left after weights and sharded optimizer state, 8k is the practical line.

## Why Qwen reaches 131k on this same trainer

`tasktrove_dq_sweep_30b.yaml` runs 131,072-token context with `attn_backend: sdpa`, **`flash_attn: false`**, and **`context_parallel_size: 1`**. It is not doing anything exotic — SDPA never materializes `seq × seq`, so attention is O(s). CP=2 appears only in the `_cp6` variant and was introduced for a Megatron backward-pass OOM in the expert path, not for attention.

The difference is that Qwen's attention is **uniformly causal**, so `is_causal=True` covers every layer with no mask tensor.

## Why SDPA alone does not close this for Grug

`grug_moe.py:599`:

```python
is_long = layer_idx % 4 == 3 or layer_idx == len(self.layers) - 1
```

Grug uses interleaved local–global attention. Of its 26 layers:

- **7 are full causal** (indices 3, 7, 11, 15, 19, 23, 25) — SDPA `is_causal=True` solves these outright.
- **19 are sliding-window** — SDPA has no native window argument. Expressing the window means a dense `attn_mask`, which materializes `seq × seq` again.

Since gradient checkpointing makes peak memory the *worst single layer*, fixing 7 of 26 leaves the ceiling unchanged.

What covers the windowed layers is either **FlashAttention-2's native `window_size`**, which does it in one argument, or **FlexAttention's `mask_mod`**.

## The question

**Why is FlashAttention rejected for Grug?** It has native sliding-window support and would close this directly. Is `_supports_flash_attn = False` (and `_supports_sdpa = False`) a correctness stance, or simply not-yet-validated? That determines whether this is a validation task or a design change.

Two smaller ones:

- Was the FP32 attention chosen for numerical stability, or is it incidental to matching the Levanter reference? BF16 would halve every score tensor — worth ~2×, moving 8k to ~11k. Not a solution, but nearly free if it is safe.
- What context length does Snowball RL actually need? If 4–8k suffices this is a documentation gap, and the limit belongs in `docs/grug-fsdp2-training.md` next to the other constraints.

Context: marin-community/marin#7786 tracks the Snowball RL campaign; #110 established this path and its constraints; #231 raises the separate AdamH question.

@yonromai flagging you as the author of #110.

**Comment by yonromai** [comment]
Looking into it; seems like I'll need to add support for FlashAttention
